fix: increment and refactor number of hops for routed messages

stedfn opened the pull request.

saketh-are reviewed (approved):
Thanks!

Could be worth adding a couple of basic unit tests covering the 0-hop and 1-hop cases.

stedfn commented:
@saketh-are I cleaned it up a bit more and added an assert in an integration test to confirm it gets incremented. Let me know if you are happy with the changes

saketh-are commented on chain/network/src/network_protocol/network.proto:
```diff
@@ -373,11 +373,14 @@ message SignedTransaction {
   google.protobuf.Timestamp created_at = 2;
   // Number of peers this routed message travelled through. Doesn't include the peer that created the message.
-  optional int32 num_hops = 3;
+  uint32 num_hops = 4;
```
Just want to confirm, are you completely sure that deprecating this field and reusing the field name is OK?

stedfn replied:
It should be ok:
- reusing the field name is fine as the protobuf encoding is based on the tag numbers
- there could be some data loss with deprecating the old field (for example if a node that uses the old version sends it to one that uses the new one) but the field was not used before anyway

There also is a protobuf backwards compatibility [test](https://github.com/near/nearcore/actions/runs/11292864089/job/31409746916?pr=12188) which passed, but I do not know how it works

saketh-are reviewed (commented):
Still LGTM, left a couple of small comments
